### DISCUSSION TITLE
How to have .optional() survive during transform()

### DISCUSSION THREAD

**Discussion by sp88011** [colinhacks/zod#3299] [Q&A] [answered]
How would I keep a property as truly optional (`{ a?:..}`) in my resulting object after doing a `.transform()`. 

Take the example below, the property is correctly inferred as optional (`a?:...`) **after** parsing **but before** transforming. 
However, after transforming, it no longer is optional.

```
const objectA = z.object({ a: z.number().optional() })

type A = z.infer<typeof objectA>;
// { a?: number  | undefined } <-- optional


const objectB = z.object({ a: z.number().optional() }).transform((x) => {
  return { a: x.a?.toString() };
});


type B = z.infer<typeof objectB>;
// { a: string  | undefined } <-- no longer optional

```

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
const objectA = z.object( { a: z.number().optional() } )

type A = z.infer<typeof objectA>
// type A = { a?: number | undefined }

const objectB = z.object( { a: z.number().optional() } ).transform( ( x ) => {
    const a = x.a?.toString()
    if ( a ) return { a }
    return {}
} )

type B = z.infer<typeof objectB>
// type B = { a: string } | { a?: undefined }
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by carlos-lopez-SaRa** [reply]
This comment saved my bacon, I had to do this to work around the issue itself, basically I had to typecast the object before merging it with the return of the transform:
![image](https://github.com/user-attachments/assets/fd7ec86d-a0d8-4ebc-8b05-d53e708ad721)

**Comment by sp88011** [comment]
Thanks! Yes, that's technically the result I want..but I'm not clear about the way to get there outside of this simple example.

I'm transforming many properties at once, some of which may be undefined. How would I do this with many properties? 

```
const objectB = z.object( {
 a: z.number().optional(),
b: z.string().optional(),
c. z.string(),
...
x: z.string().optional(),
})

.transform( ( x ) => {

// a should be {a?: undefined}
// b should be {b?: undefined}
// c should be {c: string}
//etc.

} )
```

**Reply by Crinchy** [reply]
Have you solved this issue?

**Comment by JacobWeisenburger** [comment, marked as answer]
In `zod`, an optional property means that it could be `undefined`. I understand that it is confusing, because in typescript `undefined` and optional are 2 similar things, but not the same.
